### ISSUE TITLE
Generalized management concerns (aka "Considerations")

### ISSUE THREAD

**Issue by mstenta** [farmOS/farm_grazing#22]
Currently a grazing plan has a simple textarea field that can be used to jot down "management concerns" for the plan. This is basically just a notes field for the plan, and can be used for anything that users want to store. We recently had a request to expand this to show management concerns in the rotations tab, with date ranges, and associations with one or more paddocks. This would essentially mean breaking the "management concerns" field out into a separate database table, where each row represented a separate concern (much like we are currently doing with paddock exclusions).

This would be very useful in the grazing module, especially when rendered in the rotations page. It could also be very useful in other areas of farmOS in the future, too, so perhaps this could be generalized into a core farmOS "Management Concerns" module. Similar to logs, it could allow you to add concerns and associate them with specific areas, perhaps assets, etc. Instead of a single timestamp, concerns could have a date range, or even a "date repeat" field (which is provided by the Drupal Date module and allows for storing of repeating date rules).

Since this is so similar to the current "paddock exclusions" feature we have, and since we also wanted to have the ability to use flexible/repeating dates in exclusions, it may make sense to merge the existing "paddock exclusions" feature into this new feature... so exclusions would essentially become a type of concern.

We could also generalize the way that concerns are displayed in the rotation table - so that different types of concerns could be color coded according to their type.  For instance, we discussed the possibility of having exclusion levels, to indicate "hard" exclusions (colored red) and "soft" exclusions (colored yellow).  These could just be two types of management concerns, potentially - "hard exclusions" which mean you definitely can't use the paddock, and "soft exclusions" which mean you probably shouldn't use the paddock, but maybe you will anyway.

**Comment by woodbri** [comment]
And we need to extend this to be able to include recovery periods for paddock grazing. This is simply a date range for the recovery and an association to the paddock being grazed.

**Comment by mstenta** [comment]
Mm good idea.

I like this idea of having generalized "concerns" or "considerations" that have date ranges (just as a very general concept it's nice). Maybe "Management concerns" is too specific? Or maybe it's perfect... let's think through other potential uses for this feature, and maybe a more general name will pop up that makes more sense.

**Comment by woodbri** [comment]
Currently, the rotation table is focused only on paddocks and each rotation has a record displayed in that table. To add management concerns to this table, we have two cases: 1) those linked to paddocks and 2) those without paddock links. In addition, we will be adding exclusions and recovery periods to this table.

For display purposes, I would recommend adding additional table rows to the top of the table that are not dragable for mgmt concerns that are not linked to specific paddocks. With would also be similar to the current paper charts the list mgmt concerns in the same way. For example, bird nesting, tick season, rainy season, lambing, etc. These could all be listed potential in a single row or in multiple rows (one for each concern).

The rest of the records added to this table will all be associated with one or more paddocks, can can be displayed for those paddocks like we do now.

So there are two areas for implementation:

1.  In the table generation code, we would need to add rows for the mgmt concerns and call the plot url to render the image for them. [farm_grazing.herds.rotations.inc]
2.  in the plot function, we would need to extend URI to indicate that we wanted mgmt concerns rendered.

The current URL is like: farm/grazing/plan/<plan_id>/plot/<paddock_id>/<start_str>/<end_str> or farm/grazing/plan/<plan_id>/plot/0/0/0 to generate the column header image. We could extend this to farm/grazing/plan/<plan_id>/plot/0/0/<mgmt_concern_id> and add the code to recognize that and generate the image for that mgmt concern if we want a single image or use something like id1_id2_id3_id4 etc if we want to render multiple concerns in a single image.

**Comment by mstenta** [comment]
We should also include the thoughts outlined in this issue (which I closed in favor of this more general one): https://github.com/mstenta/farm_grazing/issues/7

**Comment by mstenta** [comment]
New name: "Considerations"

**Comment by mstenta** [comment]
Jotting down an idea:

Provide two hooks that other modules can use to hook into the considerations system:

`hook_farm_grazing_consideration_types()` - Provide new consideration types that appear as options when you are adding a consideration. Each type should have a unique color for rendering in the rotation calendars. The farm_grazing module can implement this itself to provide the initial set of types (ie: 'grazing', 'recovery', 'concern', 'exclusion', and 'exclusion_urgent').

`hook_farm_grazing_considerations($paddock)` - Allow modules to provide considerations for a given paddock.  Perhaps also include date range parameters so that it can be filtered to only include considerations that fall within a given date range (like the plan's date range). The farm_grazing module can implement this itself to add the considerations it provides.

**Comment by mstenta** [comment]
In the future, we could absorb a lot of this stuff into the farm_plan module. "Plan Considerations" are something that will be useful in all plans.  The rendering component could easily be reused for crop planning, too, for example.

**Comment by mstenta** [comment]
This is done! We added a new module to farmOS called Farm Plan Considerations, which provides a database table and basic form UI for adding/editing considerations associated with a plan. In the grazing module, we migrated the exclusions to considerations, and reworked the other code for displaying them. Now there are three consideration types: "grazing exclusion", "management concern", and "urgent".  And more can be added via modules in the future.

All of the work is in the "farm_plan_considerations" branch in farmOS, and in the the "plot_considerations" branch in the grazing module.  I will merge these in now...

The only thing we didn't do from the original description is allow repeating considerations.  The date repeat module approach turned out to be too complicated to query easily, and we decided it was something that we could reapproach as a followup task. If it is requested, we can open a new issue for it in the farmOS issue queue (which is where the considerations code lives now).
